pkg/publisher/publisher.go

```diff
@@ -46,6 +46,11 @@ func (p *IPNIPublisher) Publish(ctx context.Context, providerInfo peer.AddrInfo,
 
 var _ Publisher = (*IPNIPublisher)(nil)
 
+// New creates a new IPNI publisher.
+// Thread-safety of the publisher will depend on the underlying store. If you will be calling Publish from different
```

Comment:
It's not just the store

The process is:

1. read the current head advert from the datastore
2. encode and sign a new advert using the CID of the current head as the "previous"
3. write the new advert to the datastore
4. write the new head advert CID to the datastore

See https://github.com/storacha/ipni-publisher/blob/f59f9fc3b0b8596d92dcc2b7f8e36d09417838ce/pkg/publisher/publisher.go#L184-L244

So if we try to publish another advert after 1 but before 4 then we can lose an advert. There needs to be a write lock around `Publish` or more likely just these setps within the process.

Reply:
you are totally right, sorry for missing that 🤦🏻 

I see the very first thing `Publish` does is handling chunk links, even before reading the head advert. I assume those operations also need to be synchronised, since we first fetch the chunk link for that `provider/contextID` and decide whether we need to put/delete/do nothing depending on that.

If that's the case, I'd protect the whole `Publish` method, I think having different critical regions adds little value. And since it's a common design principle in Go to not impose synchronisation by default and let the user of the type decide whether they need it or not or how they want to implement it, I'd leave the code as is, warning the user about the method not being thread-safe.

WDYT? If you agree with the above, I'll revert my previous changes and add a comment to `Publish` to indicate it's not thread-safe.